### PR TITLE
Proxy defaults

### PR DESCRIPTION
**Pull request by JensOgorek**
Handle default properties for proxies on creation without expander esp being ready.

This is a followup for #18

The defaults are handled by a static function in the module class.

- [x] create functionality to get defaults
- [x] handle new creation of properties
- [x] handle properties in proxy
- [x] change input & output module
- [x] test current version
- [x] add all modules

### CODE REVIEW

File: main/modules/module.cpp [hunk 1]
```diff
@@ -419,3 +419,12 @@ void Module::write_property(const std::string property_name, const ConstExpressi
...
 }
+
+std::map<std::string, Variable_ptr> Module::get_default_properties_for_type(const std::string &module_type) {
+    if (module_type == "Output")
+        return Output::default_properties();
+    if (module_type == "Input")
+        return Input::default_properties();
+    // We'll add other module types as we implement them
+    return {};
+}
```
Comment (falkoschindler) [inline main/modules/module.cpp:430]: @JensOgorek Can't we avoid the need for countless if-conditions and use inheritance instead?

**Comment by JensOgorek** [comment]
Replaced the static default_properties() methods with a virtual get_default_properties() function in the Module class, allowing each module type to define its own properties while eliminating the need for a separate list to maintain.

File: main/modules/output.cpp [hunk 2]
```diff
@@ -3,9 +3,14 @@
...
+std::map<std::string, Variable_ptr> Output::get_default_properties() const {
+    return {
+        {"level", std::make_shared<IntegerVariable>(-1)},
+        {"change", std::make_shared<IntegerVariable>(-1)}};
+}
+
 Output::Output(const std::string name) : Module(output, name) {
-    this->properties["level"] = std::make_shared<IntegerVariable>();
-    this->properties["change"] = std::make_shared<IntegerVariable>();
+    this->properties = this->get_default_properties();
```
Comment (falkoschindler) [inline main/modules/output.cpp:13]: @JensOgorek I think we can do that in the `Module` base class. Then individual modules only need to overwrite `get_default_properties` and are done.

**Review by falkoschindler — CHANGES_REQUESTED** [review]
As we just discussed, we should keep the initial "level" and "change" values 0.

Apart from that, we can't call `Module.create` on the core ESP. We need a different solution.

**Comment by falkoschindler** [comment]
@JensOgorek Whatever you did, somehow this PR now also contains changes from main which makes it hard to review. 😕

**Comment by JensOgorek** [comment]
> @JensOgorek Whatever you did, somehow this PR now also contains changes from main which makes it hard to review. 😕

I got it fixed. Somehow, the main > proxy_defaults merge failed.

**Review by falkoschindler — CHANGES_REQUESTED** [review]
There are still changes from main contained in this PR, like the whole "check_strapping_pins" topic in expander.cpp.

Maybe we need to start over with a fresh branch, cherry-pick individual commits and delete and burn this branch afterwards... 🔥

**Comment by JensOgorek** [comment]
> There are still changes from main contained in this PR, like the whole "check_strapping_pins" topic in expander.cpp.
> 
> Maybe we need to start over with a fresh branch, cherry-pick individual commits and delete and burn this branch afterwards... 🔥

Good idea, I bring the flamethrower tomorrow

**Comment by JensOgorek** [comment]
Flamethrower was active, and we have a new pull request with #104

**Pull request closed without merging**